### ISSUE TITLE
CosineAnnealingWarmRestarts Still Follows PyTorch 1.0 Step Order

### ISSUE THREAD

**Issue by LLYX** [pytorch/pytorch#23480]
## 🐛 Bug

Although the official step order seems to have changed to calling scheduler.step() after each iteration/epoch instead of before, this causes CosineAnnealingWarmRestarts to have an extraneous step at iteration/epoch 0, instead of starting at the maximum specified learning rate. 

## To Reproduce

1. Initialize a scheduler with T_0=1, T_mult=2.
1. Loop through using a for loop, first printing the lr, then stepping.

>>> scheduler = CosineAnnealingWarmRestarts(optimizer, T_0=1, T_mult=2)
>>> for i in range(9):
...          print(i)
...          print(scheduler.get_lr())
...          scheduler.step()
...
0
[0.0]
1
[0.5]
2
[0.5]
3
[0.25]
4
[0.5]
5
[0.42677669529663687]
6
[0.25]
7
[0.07322330470336313]
8
[0.5]

## Expected behavior

For cases such as the one above, would expect the restart to occur at the 8th iteration/epoch (index 7 above), but instead occurs at the 9th iteration/epoch (index 8 above). This is probably due to the resetting of last_epoch in the init, since it resets the last_epoch to the initial user given parameter value, despite the fact that the call to the super() init sets it to user given parameter value + 1.

## Environment

 - PyTorch Version (e.g., 1.0): 1.1
 - OS (e.g., Linux): Windows/WSL
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source):
 - Python version: 3.7
 - CUDA/cuDNN version: 10
 - GPU models and configuration: GTX 1070
 - Any other relevant information: N/A

## Additional context

N/A

**Comment by vincentqb** [comment]
Also mentioned [here](https://github.com/pytorch/pytorch/pull/21800#issuecomment-514697869).

**Comment by colesbury** [comment]
@vincentqb does your PR address this?

**Comment by gchanan** [comment]
optimistically assigning @vincentqb based on the PR linked; if that doesn't fix it, please unassign yourself.

**Comment by vincentqb** [comment]
> optimistically assigning @vincentqb based on the PR linked; if that doesn't fix it, please unassign yourself.

No, the linked PR does not addressed this. I linked only for reference.

**Comment by vincentqb** [comment]
As discussed in #21800, `get_lr` should not be used to obtain the current learning rate, as it implicitly takes an extra step.
